src/awkward/_v2/contents/indexedarray.py

```diff
@@ -998,13 +998,19 @@ def _reduce_next(
             keepdims,
         )
 
+        # If we are reducing the contents of this layout,
+        # then we do NOT want to return an optional layout
         if not branch and negaxis == depth:
             return out
         else:
             if isinstance(out, ak._v2.contents.RegularArray):
                 out = out.toListOffsetArray64(True)
 
-            elif isinstance(out, ak._v2.contents.ListOffsetArray):
+            # If the result of `_reduce_next` is a list, and we're not applying at this
+            # depth, then it will have offsets given by the boundaries in parents.
+            # This means that we need to look at the _contents_ to which the `outindex`
+            # belongs to add the new index
+            if isinstance(out, ak._v2.contents.ListOffsetArray):
```

Comment:
How are you sure that this needs to be an `if` and not an `elif`?

We have no tests that satisfy the `if isinstance(out, ak._v2.contents.RegularArray)` predicate. (I put an Exception in there and ran the tests: none of them failed.)

Aha: it's what happened in v1:

https://github.com/scikit-hep/awkward-1.0/blob/34094aa26e498210047428b2ecde6d0b1aa1fc7f/src/libawkward/array/IndexedArray.cpp#L2260-L2265

Although it would be better to have tests, conformance with v1 is a winning argument.

Reply:
Also, it's what we have in IndexedOptionArray, and that's very similar to the IndexedArray case. In fact, the IndexedArray _might_ be superfluous, since they had to share an implementation in v1, when IndexedOptionArray was a templated variation on IndexedArray. So this _might_ be a historical artifact.